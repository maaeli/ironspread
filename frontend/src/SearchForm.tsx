import React, {
  FunctionComponent,
  ReactChild,
  ReactChildren,
  ReactElement,
} from 'react';

type InputWithLabelProps = {
  id: string;
  label?: string;
  type?: string;
  value: string;
  isFocused: boolean;
  onInputChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  children?: ReactChild | ReactChildren;
};

const InputWithLabel: FunctionComponent<InputWithLabelProps> = ({
  id,
  label,
  type = 'text',
  value,
  isFocused,
  onInputChange,
  children,
}: InputWithLabelProps): ReactElement => {
  const inputRef = React.useRef<HTMLInputElement>(null);
  React.useEffect(() => {
    if (isFocused && inputRef.current) {
      inputRef.current.focus();
    }
  }, [isFocused]);
  return (
    <>
      <label htmlFor={id} className="label">
        {children}
      </label>
      &nbsp;
      <input
        ref={inputRef}
        id={id}
        type={type}
        value={value}
        autoFocus={isFocused}
        onChange={onInputChange}
        className="input"
      />
    </>
  );
};

type SearchFormProps = {
  searchTerm: string;
  onSearchInput: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onSearchSubmit: (event: React.FormEvent) => void;
};

const SearchForm: FunctionComponent<SearchFormProps> = ({
  searchTerm,
  onSearchInput,
  onSearchSubmit,
}: SearchFormProps): ReactElement => (
  <>
    <form onSubmit={onSearchSubmit} className="search-form">
      <InputWithLabel
        id="search"
        value={searchTerm as string}
        isFocused
        onInputChange={onSearchInput}
      >
        Search:
      </InputWithLabel>
      <button
        type="submit"
        disabled={!searchTerm}
        className="button button_large"
      >
        Submit
      </button>
    </form>
    <p>
      Searching for <strong>{searchTerm}</strong>
    </p>
  </>
);

export default SearchForm;
export { InputWithLabel, InputWithLabelProps, SearchFormProps };
