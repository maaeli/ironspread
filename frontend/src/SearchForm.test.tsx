import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import SearchForm from './SearchForm';

describe('SearchForm', () => {
  const TestSearchFormProps = {
    searchTerm: 'React',
    onSearchInput: jest.fn(),
    onSearchSubmit: jest.fn(),
  };

  test('renders the input field with its value', () => {
    render(<SearchForm {...TestSearchFormProps} />);
    expect(screen.getByDisplayValue('React')).toBeInTheDocument();
  });

  test('renders the correct label', () => {
    render(<SearchForm {...TestSearchFormProps} />);
    expect(screen.getByLabelText(/Search/)).toBeInTheDocument();
  });

  test('calls onSearchInput on input field change', () => {
    render(<SearchForm {...TestSearchFormProps} />);
    fireEvent.change(screen.getByDisplayValue('React'), {
      target: { value: 'Redux' },
    });
    expect(TestSearchFormProps.onSearchInput).toHaveBeenCalledTimes(1);
  });

  test('calls onSearchSubmit on button submit click', () => {
    render(<SearchForm {...TestSearchFormProps} />);
    fireEvent.submit(screen.getByRole('button'));
    expect(TestSearchFormProps.onSearchSubmit).toHaveBeenCalledTimes(1);
  });
});
