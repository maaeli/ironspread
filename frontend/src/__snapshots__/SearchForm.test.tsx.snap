// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm renders snapshot 1`] = `
<form
  class="search-form"
>
  <label
    class="label"
    for="search"
  >
    Search:
  </label>
  Â 
  <input
    class="input"
    id="search"
    type="text"
    value="React"
  />
  <button
    class="button button_large"
    type="submit"
  >
    Submit
  </button>
</form>
`;
